### ISSUE TITLE
[Feature request] Exclusive prefix sum, `torch.cumsum(input, dim=0, exclusive=True)`

### ISSUE THREAD

**Issue by yueyericardo** [pytorch/pytorch#76191]
### 🚀 The feature, motivation and pitch

Currently `torch.cumsum` only supports inclusive prefix sum. However, the exclusive prefix sum is also needed in some cases.

Proposed to add a `exclusive` parameter to the `torch.cumsum` function
```py
input = torch.tensor([1, 1, 1])
torch.cumsum(input, dim=0, exclusive=True)   # [1, 2, 3]
torch.cumsum(input, dim=0, exclusive=False)  # [0, 1, 2]
```

Related issue from PyTorch Forums:
- [PyTorch equivalent of exclusive cumsum? - PyTorch Forums](https://discuss.pytorch.org/t/pytorch-equivalent-of-exclusive-cumsum/107259)

CUDA implementation from CUB:
- [CUB: cub::DeviceScan Struct Reference](https://nvlabs.github.io/cub/structcub_1_1_device_scan.html#a02b2d2e98f89f80813460f6a6ea1692b)

**Comment by ngimel** [comment]
While we are at it, we also might add "full" that produces a tensor that's one larger than input with "0" at the start and inclusive some in the end, it's used sometimes. 
THe problem with that is that cub functions blow up binary size and compile time, they are one of biggest offenders, and each of these functions will require a slightly different cub kernel compiled.

**Comment by thomasw21** [comment]
> While we are at it, we also might add "full" that produces a tensor that's one larger than input with "0" at the start and inclusive some in the end, it's used sometimes.

Constructing this tensor in one go would be incredible

**Comment by ezyang** [comment]
Bumping the priority as internally in FB we have versions of these operators and we really shouldn't.

**Comment by ezyang** [comment]
Maybe we can use Triton to conveniently get all the variants.

**Comment by esoha-nvidia** [comment]
> > While we are at it, we also might add "full" that produces a tensor that's one larger than input with "0" at the start and inclusive some in the end, it's used sometimes.
> 
> Constructing this tensor in one go would be incredible

The best option, in my opinion, is a full sum that returns the total and let the user then narrow it as needed.  This covers all the cases.  I've written one that is working pretty okay for me.

```cuda
#include <pybind11/pybind11.h>
#include <torch/extension.h>
#include <cub/cub.cuh>

template <typename T>
T full_cumsum_helper(T *input, int64_t input_size, T *output, torch::ScalarType input_scalar_type) {
  void *d_temp_storage = nullptr;
  size_t temp_storage_bytes = 0;
  cub::DeviceScan::ExclusiveSum(d_temp_storage, temp_storage_bytes, input, output, input_size);

  // Make temp storage on in the torch mempool.
  auto temp_storage_tensor_options = torch::TensorOptions().dtype(torch::kUInt8).device(torch::kCUDA);
  auto temp_storage_tensor = torch::empty({int64_t(temp_storage_bytes)}, temp_storage_tensor_options);
  assert(temp_storage_tensor.nbytes() >= temp_storage_bytes);
  d_temp_storage = temp_storage_tensor.data_ptr<uint8_t>();

  cub::DeviceScan::ExclusiveSum(d_temp_storage, temp_storage_bytes, input, output, input_size);

  // cumsum is now in the output but the final value is missing.

  // Make storage in pinned_memory.
  auto total_tensor_options = torch::TensorOptions().dtype(input_scalar_type).pinned_memory(true);
  auto total_tensor = torch::empty({int64_t(1)}, total_tensor_options);

  // Now we run a kernel that will complete the full_cumsum.  The
  // final element of input gets added to the penultimate element of
  // output and written into the final element of output.  It also
  // gets written into total_tensor.
  complete_cumsum<<<1,1>>>(&input[input_size-1],
                           &output[input_size-1],
                           total_tensor.data_ptr<T>());
  // complete_cumsum must finish before we can read the total_tensor.
  cudaDeviceSynchronize();
  return *total_tensor.data_ptr<T>();
}

// full cumsum and also returns the total to the CPU.
int64_t full_cumsum(const torch::Tensor input, const torch::Tensor output) {
  assert(input.size(0) + 1 == output.size(0));
  switch (input.scalar_type()) {
    case torch::kInt32:
      return full_cumsum_helper(input.data_ptr<int32_t>(), input.size(0),
                                output.data_ptr<int32_t>(), input.scalar_type());
      break;
    default:
      // TODO, add all the other cases.
      std::cerr << "Not yet implemented." << std::endl;
      assert(0);
      return 0;
      break;
  }
}

PYBIND11_MODULE(TORCH_EXTENSION_NAME, m) {
  m.def("full_cumsum", &full_cumsum, "blah blah");
}
```

and in the python file:

```python
from torch.utils.cpp_extension import load

make_key_cuda = load(
  name='cumsum',
  sources=["cumsum.cu"],
  build_directory='.',
  verbose=True
)

is_valid_scan = torch.empty([is_valid.size(0)+1], dtype=is_valid.dtype, device=is_valid.device)
is_valid_total = cumsum.full_cumsum(is_valid, is_valid_scan)
```

`is_valid_scan` will have the full cumulative sum of `is_valid`.  You can use `torch.narrow` on this to get the inclusive or exclusive scan by narrowing to remove either the first or last element.  `is_valid_total` uses pinned memory to put the total into the CPU so if you need the total on CPU, use that.  If you need it on GPU, access the last element of `is_valid_scan`, which will contain the same value but on GPU memory.  If you don't need the sum on the CPU then you can remove the synchronization and clean up that part of the code.

**Comment by davidberard98** [comment]
@ezyang @albanD @soulitzer any opinions on the name of the `full` argument to prepend 0 to the cumsum result?

(context: for NestedTensor, whenever you need to convert lengths to offsets, you want to do a cumsum that's prefixed with a 0. Right now this is just annoying for testing, but I imagine that eventually there will be real NT-related use cases that would like a convenient way to do this)

**Comment by ezyang** [comment]
Bah, the annoyance is you can't just add both full and exclusive, because they are mutually exclusive.

Claude hallucinated prepend/append as the Numpy API. The defaults here would be prepend=False, append=True; exclusive is prepend=True, append=False, and full is prepend=True, append=True.

**Comment by davidberard98** [comment]
@ezyang wdyt about just adding `prepend` and users can do a view if they want just the exclusive version?

**Comment by ezyang** [comment]
Seems fine ig lol

**Comment by kwen2501** [comment]
Curious, what is the status of this request?

**Comment by davidberard98** [comment]
Sorry, I got halfway through implementing it and got busy - I probably won't be working on this in the near future

https://github.com/pytorch/pytorch/pull/127675 and https://github.com/pytorch/pytorch/pull/128469 are partial implementations (really old), but I don't think I ever got to the hard part of actually writing the cuda implementations
